rust/test/default.nix

```diff
@@ -1,12 +1,70 @@
 { holonix, pkgs }:
 let
-  name = "hc-rust-test";
-
-  script = pkgs.writeShellScriptBin name
+  hc-rust-test = pkgs.writeShellScriptBin "hc-rust-test"
   ''
   hc-rust-wasm-compile && HC_SIMPLE_LOGGER_MUTE=1 RUST_BACKTRACE=1 cargo test --all "$1" -- --test-threads=${holonix.rust.test.threads};
   '';
+
+  hc-rust-coverage-kcov = pkgs.writeShellScriptBin "hc-rust-coverage-kcov"
+  ''
+  # we need the kcov tool and curl for fetching the codecov.io uploader
+  nix-env -f https://github.com/NixOS/nixpkgs-channels/archive/nixos-19.09.tar.gz -iA kcov curl
+
+  # using cargo-make for executing kcov
+  cargo install cargo-make
+
+  # some tests require compiled wasm
+  hc-rust-wasm-compile
+
+  # kcov does not work with the global /holochain-rust/target
+  mkdir -p target
+
+  # actually kcov does not work with workspace target either
+  # we need to use targets in each crate - but that is slow
+  # use symlinks so we don't have to recompile deps over and over
+  for i in ''$(find crates -maxdepth 1 -mindepth 1 -type d | sort); do
+
+    # delete all other test binaries so they don't get run multiple times
+    rm -rf $(find target/debug -maxdepth 1 -mindepth 1 -type f)
+
+    echo "-------"
+    echo "coverage for '$i'"
+    echo "-------"
+
+    # cd into crate dirs
+    # remove any pre-existing target dir
+    # symlink to shared target dir
+    # build the test binaries
+    # run the code coverage
+
+    ( \
+      cd $i && \
+      rm -rf target || true && \
+      ln -s ../../target
+      export CARGO_TARGET_DIR=$(readlink -f ./target) && \
```

Comment:
cargo make / kcov are very particular about the placement of the target directory - this breaks incremental builds so the coverage run takes 46 minutes - but I wasn't able to get it to work in any other way.